File: 240102/돌 잘 치우기/clear-stones-well.cpp
#include <iostream>
#include <queue>
#include <algorithm>
#define MAX_N 100

using namespace std;

int n, k, m;    // n: 격자 크기, k: 시작점 개수, m: 치울 돌 개수
queue<pair<int, int> > q;

int grid[MAX_N][MAX_N];             // 격자
bool visited[MAX_N][MAX_N];         // 방문 확인 배열
int start[MAX_N*MAX_N][2];          // 모든 시작점
vector<pair<int, int> > stones;     // 돌이 들어있는 좌표들 저장
vector<pair<int, int> > removed;     // 현재 턴에서 치워야할 돌
int ans;
int different_stone;
bool InRange(int x, int y){
    return 0 <= x && x < n && 0 <= y && y < n;
}

bool CanGo(int x, int y){
    return InRange(x,y) && !grid[x][y] && !visited[x][y];
}

void do_remove(){
    for(int i=0; i<(int)removed.size(); i++){
        int cx = removed[i].first, cy = removed[i].second;
        grid[cx][cy] = 0;
    }
}

void do_recover(){
    for(int i=0; i<(int)removed.size(); i++){
        int cx = removed[i].first, cy = removed[i].second;
        grid[cx][cy] = 1;
    }
}

void initialize(){
    // 방문 배열 초기화
    for(int i=0; i<n; i++)
        for(int j=0; j<n; j++)
            visited[i][j] = false;

    // 시작점들 큐에 넣기
    for(int i=0; i<k; i++){
        int sx = start[i][0], sy = start[i][1];
        visited[sx][sy] = true;
        q.push(make_pair(sx, sy));
    }
    // 현재 턴에서의 갈 수 있는 서로 다른 돌 개수 초기화
    different_stone = k;
}

void bfs(){
    int dirs[4][2] = {{1,0},{0,1},{-1,0},{0,-1}};

    while(!q.empty()){
        pair<int,int> curr = q.front();
        q.pop();
        for(int d=0; d<4; d++){
            int nx = curr.first+dirs[d][0];
            int ny = curr.second+dirs[d][1];
            if(CanGo(nx,ny)){
                different_stone++;
                visited[nx][ny] = true;
                q.push(make_pair(nx,ny));
            }
        }
    }

    // 탐색이 끝난후 서로 다른 칸의 최댓값 갱신
    ans = max(ans, different_stone);
}

void Move(){
    initialize();
    bfs();
}


void choose_stone(int t, int index){   // t번째 돌을 고르는 함수, index는 현재 stones 벡터에서 선택할 시작
    if(t == m){
        // 만약 m개를 다 골랐다면
        // 그 돌들 치우고 bfs 탐색하여 이동한 칸 수 세기
        do_remove();
        Move();
        do_recover();
        return;
    }

    for(int i=index; i<(int) stones.size(); i++){
        // 돌을 하나 선택
        removed.push_back(make_pair(stones[i].first, stones[i].second));
        // 다음 돌 선택
        choose_stone(t+1, i+1);
        // 선택 취소
        removed.pop_back();
    }
}

int main() {
    // 입력:
    cin >> n >> k >> m;
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            cin >> grid[i][j];  // 0: 이동 가능, 1: 이동X
            if(grid[i][j])
                stones.push_back(make_pair(i,j));   // 만약 돌이 있다면 좌표 저장
        }
    }
    // 시작점 저장
    for(int i=0; i<k; i++){
        cin >> start[i][0] >> start[i][1];
        start[i][0]--; start[i][1]--;
    }

    // stones 중에서 m개 돌을 backtracking으로 고르기
    choose_stone(0,0);    // 0번째 stone 고르는 함수
    
    // 출력
    cout << ans;
    return 0;
}

File: 240127/돌 잘 치우기/clear-stones-well.py
from collections import deque

# 변수 선언 및 입력
n, k, m = tuple(map(int, input().split())) # n: 격자 크기, k: 시작점 개수, m: 치워야 할 돌 개수
grid = [list(map(int, input().split())) for _ in range(n)]  # 0: 빈칸, 1: 돌

ans = 0
start_points = []
stones = [ (i,j) for i in range(n) for j in range(n) if grid[i][j] ]
selected_stones = []

# bfs에 필요한 변수들
q = deque()
visited = [[False for _ in range(n)] for _ in range(n)]

def InRange(x, y):
    return 0 <= x and x < n and 0 <= y and y < n

def CanGo(x, y):
    return InRange(x,y) and not visited[x][y] and not grid[x][y]

def Initialize_visited():
    for i in range(n):
        for j in range(n):
            visited[i][j] = False
    
def bfs():
    # queue에 남은 것이 없을때까지 반복합니다.
    while q:
        # queue에서 가장 먼저 들어온 원소를 뺍니다.
        x, y = q.popleft()

        dxs = [-1,1,0,0]
        dys = [0,0,-1,1]
        # queue에서 뺀 원소의 위치를 기준으로 4방향을 확인해봅니다.
        for dx, dy in zip(dxs, dys):
            nx, ny = x+dx, y+dy

            # 아직 방문한 적이 없으면서 갈 수 있는 곳이라면
            # 새로 queue에 넣어주고 방문 여부를 표시해줍니다. 
            if CanGo(nx,ny):
                visited[nx][ny] = True
                q.append((nx,ny))

def Move():
    global ans
    Initialize_visited()
    # 돌 치우기
    for s in selected_stones:
        grid[s[0]][s[1]] = 0
    
    # k개의 시작점을 queue에 넣고 시작합니다.
	# BFS는 여러 시작점에서 시작하여
    # 이동 가능한 칸을 전부 탐색하는 것이 가능합니다.
    for start in start_points:
        visited[start[0]][start[1]] = True
        q.append(start)
    
    bfs()

    curr = sum([1 for i in range(n) for j in range(n) if visited[i][j]])
    ans = max(curr, ans)
    
    # 돌 원래대로
    for s in selected_stones:
        grid[s[0]][s[1]] = 1


# m개의 돌 조합 구하기
def Choose(idx, cnt):       # idx는 현재 stones 에서 선택할 시작점, cnt번째 돌을 고르는 함
    if cnt == m:
        Move()
        return
    
    for i in range(idx, len(stones)):
        selected_stones.append(stones[idx])
        Choose(idx+1, cnt+1)
        selected_stones.pop()

# 시작점 입력받기
for _ in range(k):
    r, c = tuple(map(int, input().split()))
    start_points.append((r-1,c-1))

Choose(0,0)
print(ans)